File: cogs/groups.py
import logging
from typing import cast

import discord
from discord import app_commands
from discord.ext import commands

from core import config
from core.issues import BadGroupIssueModal, report_bad_group
from core.models import WoWPlayer
from services.session_service import SessionService

logger = logging.getLogger(__name__)


class Groups(commands.Cog):
    def __init__(self, bot: commands.Bot):
        self.bot = bot
        self.session_service = SessionService(bot)

    @commands.hybrid_command(name="wheel")
    async def wheel(self, ctx: commands.Context[commands.Bot]) -> None:
        """Generates a series of embed messages that shows groups of players split into 5 person teams."""
        await ctx.defer()
        try:
            if hasattr(self.bot, "group_service"):
                await self.bot.group_service.core_wheel(ctx, debug_value=False)
            else:
                await ctx.send("❌ GroupService not initialized.")
        except discord.HTTPException as e:
            await ctx.send(f"❌ Discord API Error: {e.status} - {e.text}")
        except Exception as e:
            await ctx.send("❌ An unexpected error occurred. Please try again later.")
            logger.error("Error in wheel command: %s", e)

    @commands.hybrid_command(name="newwheel")
    async def newwheel(self, ctx: commands.Context[commands.Bot]) -> None:
        """Generates groups with an enhanced rolling reveal animation and sound effects."""
        await ctx.defer()
        try:
            if hasattr(self.bot, "group_service"):
                await self.bot.group_service.core_wheel(
                    ctx, debug_value=False, enhanced=True
                )
            else:
                await ctx.send("❌ GroupService not initialized.")
        except discord.HTTPException as e:
            await ctx.send(f"❌ Discord API Error: {e.status} - {e.text}")
        except Exception as e:
            await ctx.send("❌ An unexpected error occurred. Please try again later.")
            logger.error("Error in newwheel command: %s", e)

    @commands.hybrid_command(name="activity")
    async def activity(self, ctx: commands.Context[commands.Bot]) -> None:
        """Starts an enhanced wheel and creates a Discord Activity invite for the voice channel."""
        await self._execute_activity(ctx, debug=False)

    @commands.hybrid_command(name="wheelson")
    async def wheelson(self, ctx: commands.Context[commands.Bot]) -> None:
        """Starts an enhanced wheel and creates a Discord Activity invite for the voice channel."""
        await self._execute_activity(ctx, debug=False)

    @commands.command(name="activitytest", hidden=True)
    async def activitytest(self, ctx: commands.Context[commands.Bot]) -> None:
        """Hidden prefix command to test the activity with debug data."""
        await self._execute_activity(ctx, debug=True)

    async def _execute_activity(
        self, ctx: commands.Context[commands.Bot], debug: bool
    ) -> None:
        """Internal helper to execute the activity logic with optional debug data."""
        await ctx.defer()
        try:
            if not hasattr(self.bot, "group_service"):
                await ctx.send("❌ GroupService not initialized.")
                return

            # Check if user is in voice
            if not (ctx.author.voice and ctx.author.voice.channel):
                await ctx.send(
                    "❌ You must be in a voice channel to start an activity."
                )
                return

            # Use get_groups_data merely to validate and get initial players.
            # We don't need the groups calculated yet.
            result = await self.bot.group_service.get_groups_data(ctx, debug=debug)
            if not result:
                return  # Error message already sent

            players = cast(list[WoWPlayer], result["players"])

            # Create Session in Firebase
            session_id = await self.session_service.create_session(
                ctx, players, debug=debug
            )

            if not session_id:
                await ctx.send("❌ Failed to create session. Is Firebase configured?")
                return

            # Create Links
            channel = ctx.author.voice.channel
            app_id = config.DISCORD_APPLICATION_ID

            # Create Embedded Invite
            invite_url = "N/A"
            if app_id:
                invite = await channel.create_invite(
                    target_type=discord.InviteTarget.embedded_application,
                    target_application_id=int(app_id),
                    max_age=300,  # 5 minutes
                )
                invite_url = invite.url

            # Create Direct Link
            activity_url_base = config.ACTIVITY_URL

            msg = "🎮 **Join the Activity!**\n"
            msg += f"**Voice Channel Activity:** {invite_url}\n"

            if activity_url_base:
                direct_link = f"{activity_url_base}?sessionId={session_id}"
                msg += f"**Browser Link:** [Click Here]({direct_link})\n"
            else:
                msg += "⚠️ `ACTIVITY_URL` not set in .env."

            await ctx.send(msg)

        except discord.HTTPException as e:
            await ctx.send(f"❌ Discord API Error: {e.status} - {e.text}")
        except Exception as e:
            await ctx.send("❌ An unexpected error occurred. Please try again later.")
            logger.error("Error in activity command: %s", e)

    @commands.Cog.listener()
    async def on_voice_state_update(
        self,
        member: discord.Member,
        before: discord.VoiceState,
        after: discord.VoiceState,
    ) -> None:
        """
        Listen for voice state updates to sync the lobby.
        """
        # We only care if someone joined or left a channel that has an active session
        channel = after.channel or before.channel
        if not channel:
            return

        # Check if we have a session for this channel
        if channel.id in self.session_service.active_sessions:
            members = [m for m in channel.members if not m.bot]
            # Update lobby
            await self.session_service.update_lobby_players(channel.id, members)

    @commands.hybrid_command(
        name="badgroup", description="Report a bad set of groups to the developers."
    )
    @app_commands.describe(
        title="What's the issue? (e.g. Too many melee in Group 1)",
        description="Please explain why these groups are not ideal...",
    )
    async def badgroup(
        self,
        ctx: commands.Context[commands.Bot],
        title: str | None = None,
        *,
        description: str | None = None,
    ):
        """
        Report a bad set of groups to the developers.
        NOTE: Bad group reports MUST ALWAYS include logs for debugging purposes.
        Do not add a toggle to disable log inclusion for this command.
        """
        if not hasattr(self.bot, "group_service"):
            await ctx.send("❌ GroupService not initialized.", ephemeral=True)
            return

        guild_id = ctx.guild.id if ctx.guild else None
        if not guild_id or guild_id not in self.bot.group_service.last_results:
            await ctx.send(
                "❌ No group creation data found for this server. Run /wheel first.",
                ephemeral=True,
            )
            return

        last_results = self.bot.group_service.last_results[guild_id]

        # If called via slash command without arguments, open the modal
        if ctx.interaction and title is None:
            await ctx.interaction.response.send_modal(BadGroupIssueModal(last_results))
            return

        # If arguments are provided (either prefix or slash), process directly
        if not title or not description:
            await ctx.send(
                "❌ Please provide both a title and a description when using the prefix command. "
                'Usage: `!badgroup "Title" Description` or use `/badgroup` to open a modal.',
                ephemeral=True,
            )
            return

        # Process the report
        await ctx.defer(ephemeral=True)
        try:
            issue = await report_bad_group(ctx.author, last_results, title, description)
            await ctx.send(
                f"✅ Bad group reported successfully: {issue['html_url']}",
                ephemeral=True,
            )
        except Exception as e:
            logger.error("Error creating GitHub issue via badgroup command: %s", e)
            await ctx.send(f"❌ Failed to create issue: {e}", ephemeral=True)


async def setup(bot: commands.Bot):
    await bot.add_cog(Groups(bot))

File: packages/bot/src/commands/groups.ts
import { SessionService, type Bot, type Guild } from '../services/sessionService.js';
import type { CommandContext, GroupService } from '../services/groupService.js';
import { reportBadGroup } from '../core/issues.js';
import { ACTIVITY_URL, DISCORD_APPLICATION_ID } from '../core/config.js';
import { reportError } from '../core/sentry.js';

export interface GroupsContext {
  guild: { id: string } | null;
  author: {
    id: string;
    name: string;
    voice?: {
      channel?: {
        id: string;
        name: string;
        members?: { bot: boolean }[];
        createInvite?(): Promise<{ url: string }>;
      } | null;
    } | null;
  };
  channel?: {
    members: { bot: boolean; id: string; toString(): string }[];
    sendTyping(): Promise<void>;
  };
  send(content: string, options?: { ephemeral?: boolean }): Promise<unknown>;
  defer(options?: { ephemeral?: boolean }): Promise<void>;
  interaction?: { response: { sendModal(modal: unknown): Promise<void> } } | null;
}

export type ActivityContext = Omit<GroupsContext, 'guild'> & {
  guild: Guild | null;
};

export interface VoiceState {
  channel: { id: string; members: { bot: boolean }[] } | null;
}

export class GroupsHandler {
  bot: Bot;
  sessionService: SessionService;
  groupService: GroupService;

  constructor(bot: Bot, groupService: GroupService, sessionService?: SessionService) {
    this.bot = bot;
    this.groupService = groupService;
    this.sessionService = sessionService ?? new SessionService(bot);
  }

  /**
   * Attempts to generate a Discord invite URL for the user's current voice channel.
   * Fails gracefully if permissions, channel type, or rate limits prevent creation.
   *
   * @param voiceChannel - The voice channel object from the context
   * @returns A string containing the URL, or 'N/A' if it could not be created
   */
  private async _generateInviteUrl(
    voiceChannel: NonNullable<ActivityContext['author']['voice']>['channel'],
  ): Promise<string> {
    if (!DISCORD_APPLICATION_ID || !voiceChannel?.createInvite) {
      return 'N/A';
    }

    try {
      const invite = await voiceChannel.createInvite();
      return invite.url;
    } catch (e) {
      // Invite creation can fail for many benign reasons (missing perms,
      // channel type, rate limit). The activity link still works without
      // it; surface to Sentry but don't break the response.
      reportError(e, { tags: { handler: 'activity.createInvite' } });
      return 'N/A';
    }
  }

  /**
   * Formats the response message for the /activity command.
   *
   * @param inviteUrl - The URL to join the Discord voice channel
   * @param guildId - The ID of the Discord guild
   * @param channelId - The ID of the Discord channel
   * @returns The formatted markdown string to be sent to the channel
   */
  private _buildActivityMessage(
    inviteUrl: string,
    guildId: string,
    channelId: string,
  ): string {
    let msg = '🎮 **Join the Activity!**\n';
    msg += `**Voice Channel Activity:** ${inviteUrl}\n`;

    if (ACTIVITY_URL) {
      const directLink = `${ACTIVITY_URL}?guildId=${guildId}&channelId=${channelId}`;
      msg += `**Browser Link:** [Click Here](${directLink})\n`;
    } else {
      msg += '⚠️ `ACTIVITY_URL` not set in .env.';
    }

    return msg;
  }

  // Errors propagate to the InteractionCreate wrapper in main.ts, which
  // reports to Sentry with command/guild tags and replies with a generic
  // ephemeral error message. Catching here would prevent that.
  async wheel(ctx: GroupsContext): Promise<void> {
    await ctx.defer();
    if (!ctx.channel) {
      throw new Error('Channel context is required for coreWheel');
    }
    await this.groupService.coreWheel(ctx as unknown as CommandContext, false);
  }

  async activity(ctx: ActivityContext, debug = false): Promise<void> {
    await ctx.defer();
    if (!ctx.author.voice?.channel) {
      await ctx.send('❌ You must be in a voice channel to start an activity.');
      return;
    }

    const result = await this.sessionService.getOrCreateSession(ctx, debug);
    if (!result) {
      await ctx.send('❌ Failed to create/get session. Is Firebase configured?');
      return;
    }

    const [guildId, channelId] = result;

    const inviteUrl = await this._generateInviteUrl(ctx.author.voice.channel);
    const msg = this._buildActivityMessage(inviteUrl, guildId, channelId);

    await ctx.send(msg);
  }

  async badgroup(
    ctx: GroupsContext,
    title?: string | null,
    description?: string | null,
  ): Promise<void> {
    const guildId = ctx.guild?.id ?? null;
    const lastResults = guildId ? this.groupService.lastResults.get(guildId) : undefined;
    if (!lastResults) {
      await ctx.send(
        '❌ No group creation data found for this server. Run /wheel first.',
        { ephemeral: true },
      );
      return;
    }

    // If slash command without arguments, send modal
    if (ctx.interaction && title == null) {
      // In actual discord.js: ctx.interaction.response.sendModal(...)
      await ctx.interaction.response.sendModal(lastResults);
      return;
    }

    if (!title || !description) {
      await ctx.send(
        '❌ Please provide both a title and a description when using the prefix command. ' +
          'Usage: `!badgroup "Title" Description` or use `/badgroup` to open a modal.',
        { ephemeral: true },
      );
      return;
    }

    await ctx.defer({ ephemeral: true });
    const issue = await reportBadGroup({
      reporterName: ctx.author.name,
      reporterId: ctx.author.id,
      title,
      description,
      players: lastResults.players,
      groups: lastResults.groups,
    });
    await ctx.send(`✅ Bad group reported successfully: ${issue.html_url}`, { ephemeral: true });
  }

  async onVoiceStateUpdate(
    member: { bot: boolean; guild: Guild },
    before: VoiceState,
    after: VoiceState,
  ): Promise<void> {
    if (before.channel?.id === after.channel?.id) return;

    if (before.channel && this.sessionService.activeChannels.has(before.channel.id)) {
      const humans = before.channel.members.filter((m) => !m.bot);
      if (humans.length === 0) {
        await this.sessionService.cleanupChannel(before.channel.id);
      } else {
        await this.sessionService.updateChannelPlayers(before.channel.id, member.guild);
      }
    }

    if (after.channel && this.sessionService.activeChannels.has(after.channel.id)) {
      await this.sessionService.updateChannelPlayers(after.channel.id, member.guild);
    }
  }
}
